### PR TITLE
Implement Typography system

### PR DESCRIPTION
**Pull request by epetrow**
Related to: https://github.com/telerik/kendo-themes-private/issues/95

Design: https://www.figma.com/file/7fhpSUTqL2HAQZaWgo6WR0/Typography?type=design&node-id=0%3A1&mode=dev&t=juow950ED8qAb4lZ-1

- [x] Implement Typography Module
- [x] Integrate Typography Module in Bootstrap
- [x] Integrate Typography Module in Classic
- [x] Integrate Typography Module in Default
- [x] Integrate Typography Module in Fluent
- [x] Integrate Typography Module in Material
- [x] Integrate Typography Module in Utils

### CODE REVIEW

File: packages/core/scss/typography/index.import.scss [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+$_default-font-families: (
+    sans: $kendo-font-family-sans,
+    serif: $kendo-font-family-serif,
+    sans-serif: $kendo-font-family-sans-serif,
+    monospace: $kendo-font-family-monospace
+) !default;
+
+$kendo-font-sizes: $_default-font-sizes !default;
+
+$kendo-font-sizes: k-map-merge( $_default-font-sizes,  $kendo-font-sizes ) !default;
```
Comment (JoomFX) [inline packages/core/scss/typography/index.import.scss:88]: Remove `!default`

File: packages/core/scss/typography/index.import.scss [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+$kendo-line-heights: $_default-line-heights !default;
+
+$kendo-line-heights: k-map-merge( $_default-line-heights,  $kendo-line-heights ) !default;
```
Comment (JoomFX) [inline packages/core/scss/typography/index.import.scss:92]: Remove `!default`

File: packages/core/scss/typography/index.import.scss [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+$kendo-font-weights: $_default-font-weights !default;
+
+$kendo-font-weights: k-map-merge( $_default-font-weights,  $kendo-font-weights ) !default;
```
Comment (JoomFX) [inline packages/core/scss/typography/index.import.scss:96]: Remove `!default`

File: packages/core/scss/typography/index.import.scss [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+$kendo-letter-spacings: $_default-letter-spacings !default;
+
+$kendo-letter-spacings: k-map-merge( $_default-letter-spacings,  $kendo-letter-spacings ) !default;
```
Comment (JoomFX) [inline packages/core/scss/typography/index.import.scss:100]: Remove `!default`

File: packages/core/scss/typography/index.import.scss [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+$kendo-font-families: $_default-font-families !default;
+
+$kendo-font-families: k-map-merge( $_default-font-families,  $kendo-font-families ) !default;
```
Comment (JoomFX) [inline packages/core/scss/typography/index.import.scss:104]: Remove `!default`

File: packages/core/scss/typography/index.import.scss [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+) !default;
+
+$_default-font-families: (
+    sans: $kendo-font-family-sans,
+    serif: $kendo-font-family-serif,
+    sans-serif: $kendo-font-family-sans-serif,
+    monospace: $kendo-font-family-monospace
+) !default;
+
+$kendo-font-sizes: $_default-font-sizes !default;
```
Comment (JoomFX) [inline packages/core/scss/typography/index.import.scss:86]: A comment should be added so the Docs can pick it up:

```scss
/// The global default Font Sizes map.
/// @group typography-system
```

This should be done for all the maps and in each theme as well.

File: packages/fluent/scss/core/_index.scss [hunk 7]
```diff
@@ -46,7 +48,14 @@
     $kendo-loading-text: $kendo-loading-text !default,
     // Elevation
     $_default-elevation: $kendo-elevation,
-    $kendo-elevation: $kendo-elevation !default
+    $kendo-elevation: $kendo-elevation !default,
+    // Typography
+    $_default-font-sizes: $kendo-font-sizes,
```
Comment (JoomFX) [inline packages/fluent/scss/core/_index.scss:53]: I don't think that the private variables should be added here, but test it please.

File: packages/fluent/scss/core/_index.scss [hunk 8]
```diff
@@ -46,7 +48,14 @@
...
+    $kendo-font-sizes: $kendo-font-sizes !default,
+    $_default-line-heights: $kendo-line-heights,
```
Comment (JoomFX) [inline packages/fluent/scss/core/_index.scss:55]: ditto

**Review by JoomFX — CHANGES_REQUESTED** [review]

File: packages/core/scss/typography/index.import.scss [hunk 9]
```diff
@@ -0,0 +1,220 @@
+@import "../functions/index.import.scss";
+
+/// The base font size across all components.
+/// @group typography
+$kendo-font-size: 0.875rem !default;
+/// The extra extra small font size across all components.
+/// @group typography
+$kendo-font-size-2xs: 0.5rem !default;
```
Comment (inikolova) [inline packages/core/scss/typography/index.import.scss:8]: In relation to sizes, we use the concept for xl, xxl, xxxl -> maybe we should follow the same naming here as well

For example icons sizes - https://github.com/telerik/kendo-themes/blob/fc0520f2bd1341d64c6c0b179c30e725b3ae7e81/packages/fluent/scss/icon/_variables.scss#L13

File: packages/utils/scss/_mixins.scss [hunk 10]
```diff
@@ -8,18 +8,27 @@
...
+            $_custom-prop: if( $_key == DEFAULT, var( --#{$kendo-prefix}#{$_name}, #{$_val} ), var( --#{$kendo-prefix}#{$_name}-#{$_key}, #{$_val} ) );
 
             @if $important != only {
                 .#{$_selector} {
                     @each $prop in $_props {
-                        #{$prop}: if( $_fn, k-meta-call($_fn, $_val), $_val );
+                        @if $css-var {
+                            #{$prop}: if( $_fn, k-meta-call($_fn, $_custom-prop), $_custom-prop );
+                        } @else {
+                            #{$prop}: if( $_fn, k-meta-call($_fn, $_custom-prop), $_custom-prop );
```
Comment (inikolova) [inline packages/utils/scss/_mixins.scss:19]: The _else_ statement should not relate to custom prop

**Comment by inikolova** [comment]
As a user, I find a bit strange the different values behind the line heights in the [core package](https://github.com/telerik/kendo-themes/pull/4894/commits/6a82fa4db2f16559efc5f1459c6edb6afe9c48d9#diff-093a9842cf1081ff41b4fa9621a02aca7f3de4791e3c4ce7f5b92480560e1217R30) and [utilities](https://github.com/telerik/kendo-themes/pull/4894/commits/12cbbd51d2738b7113a0ecb148f738411d483ea7#diff-efcc058b9e5159602ad9131f90372c58fe5fda6db718648fe0f8423f27098bfbR679). If using .kendo-line-height-xs utility class in my app I would expect it gets value 1 (not 1.25) or vice-versa. If we prefer keeping the utils packages separate from the theme, I would suggest to use var names in the [utils package](https://github.com/telerik/kendo-themes/pull/4894/commits/12cbbd51d2738b7113a0ecb148f738411d483ea7#diff-efcc058b9e5159602ad9131f90372c58fe5fda6db718648fe0f8423f27098bfbR679), which differ the names in the [core](https://github.com/telerik/kendo-themes/pull/4894/commits/6a82fa4db2f16559efc5f1459c6edb6afe9c48d9#diff-093a9842cf1081ff41b4fa9621a02aca7f3de4791e3c4ce7f5b92480560e1217R118) package.

**Comment by Juveniel** [comment]
Two things:

1) Generated utility classes values are prefixed with `--k-` instead of `--kendo-`

2) Typography elements can have specific CSS variables that allow customization ( e.g --kendo-h1-font-size ).

3) Decide on using `--kendo-font-size` vs `kendo-font-size-md` and make implementation identical in all themes

File: packages/fluent/scss/index.scss [hunk 11]
```diff
@@ -342,6 +342,7 @@
     // Elevation
     @include kendo-elevation--styles();
     @include elevation--styles-legacy();
+    @include kendo-core--typography--styles();
```
Comment (inikolova) [inline packages/fluent/scss/index.scss:345]: Please move this row to the Typography section

File: packages/bootstrap/scss/dataviz/_variables.scss [hunk 12]
```diff
@@ -86,17 +86,17 @@ $kendo-series-29: $kendo-series-e-darker !default;
...
-$kendo-chart-font-size: $kendo-font-size-md !default;
-$kendo-chart-line-height: $kendo-line-height-md !default;
+$kendo-chart-font-family: var( --kendo-font-family, inherit ) !default;
+$kendo-chart-font-size: var( --kendo-font-size, inherit ) !default;
+$kendo-chart-line-height: var( --kendo-line-height, normal )!default;
 $kendo-chart-sm-font-size: 11px !default;
 $kendo-chart-md-font-size: 12px !default;
 $kendo-chart-lg-font-size: 16px !default;
-$kendo-chart-tooltip-font-size: ($kendo-font-size-md * .929) !default;
+$kendo-chart-tooltip-font-size: calc( #{$kendo-font-size-md} * .929) !default;
```
Comment (zhpenkov) [inline packages/bootstrap/scss/dataviz/_variables.scss:95]: extra `-md` suffix

File: packages/bootstrap/scss/gantt/_variables.scss [hunk 13]
```diff
@@ -1,8 +1,8 @@
 // Gantt
 $kendo-gantt-border-width: 1px !default;
-$kendo-gantt-font-family: $kendo-font-family !default;
-$kendo-gantt-font-size: $kendo-font-size-md !default;
-$kendo-gantt-line-height: $kendo-line-height-md !default;
+$kendo-gantt-font-family: var( --kendo-font-family, inherit ) !default;
+$kendo-gantt-font-size: var( --kendo-font-size, inherit ) !default;
+$kendo-gantt-line-height: var( --kendo-line-height-md, normal ) !default;
```
Comment (zhpenkov) [inline packages/bootstrap/scss/gantt/_variables.scss:5]: extra `-md` suffix

File: packages/bootstrap/scss/core/typography/index.import.scss [hunk 14]
```diff
@@ -0,0 +1,100 @@
+
+/// The base font size across all components.
+/// @group typography
+$kendo-font-size: $font-size-base !default;
+/// The extra small font size across all components.
```
Comment (inikolova) [inline packages/bootstrap/scss/core/typography/index.import.scss:5]: $kendo-font-size-xxs is missing here -> is this by purpose?

File: packages/bootstrap/scss/core/typography/index.import.scss [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+/// @group typography
+$kendo-line-height-md: $kendo-line-height !default;
+/// The large line height across all components.
+/// @group typography
+$kendo-line-height-lg: $line-height-lg !default;
+/// The base line height in ems across all components.
+/// @group typography
+$kendo-line-height-em: calc( #{$kendo-line-height} * 1em ) !default;
+
+$kendo-font-weight-light: $font-weight-light !default;
```
Comment (inikolova) [inline packages/bootstrap/scss/core/typography/index.import.scss:39]: The variables below are not documented

File: packages/bootstrap/scss/core/typography/index.import.scss (hunk 14 above)
Comment (epetrow) [reply packages/bootstrap/scss/core/typography/index.import.scss:5]: It is omitted on purpose and I am questioning whether to remove it altogethera as we are not using it anywhere

File: packages/fluent/scss/calendar/_variables.scss [hunk 16]
```diff
@@ -282,62 +282,38 @@ $kendo-calendar-end-range-hover-shadow: inset -1px -1px 0 0 if($kendo-enable-col
...
 // Calendar sizes
-/// The font size of the small Calendar.
-/// @group calendar
-$kendo-calendar-sm-font-size: $kendo-font-size-sm !default;
-/// The line height of the small Calendar.
-/// @group calendar
-$kendo-calendar-sm-line-height: $kendo-line-height-sm !default;
-/// The size of the small calendar cell.
-/// @group calendar
+$kendo-calendar-sm-font-size: var( --kendo-font-size-sm, inherit ) !default;
```
Comment (inikolova) [inline packages/fluent/scss/calendar/_variables.scss:285]: Some of the calendar variables are not documented

Comment (epetrow) [reply packages/fluent/scss/calendar/_variables.scss:285]: They actually are https://github.com/telerik/kendo-themes/blob/develop/packages/fluent/scss/calendar/_variables.scss

They have been documented as part of the unification of the Calendar

File: packages/bootstrap/scss/core/typography/index.import.scss (hunk 14 above)
Comment (inikolova) [reply packages/bootstrap/scss/core/typography/index.import.scss:5]: I would suggest to keep this value -> it is 0.5rem and is possible to use it in the future :)

**Review by zhpenkov — APPROVED** [review]

**Review by JoomFX — APPROVED** [review]

**Review by inikolova — APPROVED** [review]

**Pull request merged**